### ISSUE TITLE
ZFS doesn't play nice with virtiofs

### ISSUE THREAD

**Issue by deajan** [openzfs/zfs#14932]
I'm trying to share some ZFS datasets with a qemu guest using virtiofs.
So far, eveytime I try to launch my guest, it shows an empty folder (actually shares the mountpoint directory instead of the ZFS fs).

I had various times where it seemed like zfs would "unmount" itself once I launched my virtual guest.

After having tried the following (on a test system of course)
``` 
zfs umount -a
rm /backup/* -rf
zfs mount -a
```

I can list the my zfs dataset with:
```
zfs list
[...]
backup/dataset/mydataset     140K  34.8T      140K  /backup/dataset/mydataset
[...]
```
Content is also visible.

But I cannot use that dataset with virtiofs
```
virsh start mymachine
error: internal error: the virtiofs export directory '/backup/dataset/mydataset/' does not exist
```

I'm really sorry for the noise since this is probably a vitiofs bug and not a zfs one, but I do think zfs behaves strangely since it shouldn't just "unmount" my datasets when accessed via virtiofs. Do you guys have any experience with virtiofs ?
Do FUSE daemons get to use zfs mountpoints properly ? Anything to configure perhaps ?

Best regards

### System information
```
filename:       /lib/modules/5.14.0-284.11.1.el9_2.x86_64/extra/zfs.ko.xz
version:        2.1.11-2
license:        CDDL
author:         OpenZFS
description:    ZFS
alias:          devname:zfs
alias:          char-major-10-249
rhelversion:    9.2
srcversion:     8081FD700719F8F0FB60578
depends:        spl,znvpair,icp,zlua,zzstd,zunicode,zcommon,zavl
```

Virtiofs relevant config in virtual guest xml:
```
    <filesystem type='mount' accessmode='passthrough'>
      <driver type='virtiofs' queue='1024'/>
      <binary path='/usr/libexec/virtiofsd' xattr='on'>
        <cache mode='always'/>
      </binary>
      <source dir='/backup/dataset/mydataset/'/>
      <target dir='sometarget'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x1d' function='0x0'/>
    </filesystem>
```

**Comment by GregorKopka** [comment]
Running with 2.1.11-r0-funtoo and have no problems handing ZFS backed directories into qemu VMs, both linux and Windows.

Remember that you have to mount the shared folder in the guest!
Else you'll only see an empty mountpoint when looking in the guest, with the content being there when looking on the host.

What is your actual issue: the content _vanishing on the host_ or _not showing up in the guest_?

**Comment by deajan** [comment]
So, first, thanks for the answers, and sorry for the delay in response.
Here's what I did so far:
- Updated zfs to latest 2.1.12-1

There's no `--announce-submunts` option in libvirt, so I did an ugly patch:
- Move /usr/libexec/virtiofsd to /usr/libexec/virtiofsd_dist
- Create /usr/libexec/virtiofsd containing
```
#!/usr/bin/env bash

/usr/libexec/virtiofsd_dist $@ --announce-submounts
```

This way, I get to add `--announce-submounts` to virtiofsd:
```
# ps aux | grep virtiofsd
root     12079  0.0  0.0   6400  2328 pts/19   S+   18:36   0:00 grep --color=auto virtiof
root     63816  0.0  0.0   5776  3892 ?        S    18:23   0:00 /usr/libexec/virtiofsd --fd=98 -o source=/backup/restic_stash/,cache=always,xattr --announce-submounts
root     63818  0.0  0.0 2238128 5332 ?        Sl   18:23   0:00 /usr/libexec/virtiofsd --fd=98 -o source=/backup/restic_stash/,cache=always,xattr --announce-submounts

On the host machine, I created a fresh machine with following libvirt config
```
    <filesystem type='mount' accessmode='passthrough'>
      <driver type='virtiofs' queue='1024'/>
      <binary path='/usr/libexec/virtiofsd' xattr='on'>
        <cache mode='always'/>
      </binary>
      <source dir='/backup/restic_stash/'/>
      <target dir='restic_stash'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x1a' function='0x0'/>
    </filesystem>
```

Host config:
```
# zfs list
NAME                           USED  AVAIL     REFER  MOUNTPOINT
backup                         871G  34.4T      283G  /backup
backup/restic_stash           73.6G  34.4T      186K  /backup/restic_stash
backup/restic_stash/userA 140K  34.4T      140K  /backup/restic_stash/userA
backup/restic_stash/userB 73.6G  34.4T     73.6G  /backup/restic_stash/userB

# mount | grep zfs
backup on /backup type zfs (rw,noatime,seclabel,xattr,noacl)
backup/restic_stash on /backup/restic_stash type zfs (rw,noatime,seclabel,noxattr,noacl)
backup/restic_stash/userB on /backup/restic_stash/userB type zfs (rw,noatime,seclabel,noxattr,noacl)
backup/restic_stash/userA on /backup/restic_stash/userA type zfs (rw,noatime,seclabel,noxattr,noacl)

# touch /backup/restic_stash/restic_stash_file
# touch /backup/restic_stash/userA/userA_file
# touch /backup/restic_stash/userB/userB_file
```

Once I started the 

```
# mount | grep virtiofs
restic_stash on /restic_stash type virtiofs (rw,nosuid,nodev,noexec,noatime,nodiratime,seclabel)
# ls -alh /restic_stash/
total 512
drwx------.  2 restic root   2 Jul  7 18:18 .
dr-xr-xr-x. 20 root   root 270 May 31 23:42 ..
# umount /restic_stash
# ls -alh /restic_stash/
total 0
drwxr-xr-x.  2 root root   6 May 16 13:31 .
dr-xr-xr-x. 20 root root 270 May 31 23:42 ..

dmesg | egrep -i "virtiofs|restic_stash"
[    9.119442] systemd-fstab-generator[474]: Checking was requested for "restic_stash", but it is not a device.
[   10.535127] virtiofs virtio0: virtio_fs_setup_dax: No cache capability
```
As you can see, FS is mounted (since once I unmount it, the ls output changes slightly).
Anyway, no files in there.

Strangely enough, since zfs 2.1.12-1 update, host zfs filesystem doesn't get unmounted anymore when I start the VM.

I've also redone the above tests after disabling SELinux in host and guest systems.
Lastly, I redid the test without the `--announce-submounts` hack I tried.

**Comment by deajan** [comment]
Oh, and for what it's worth, using an xfs directory instead of a zfs one works.

**Comment by filip-paczynski** [comment]
I tested this on pop os vm :
- ZFS layout
  ```
  # zfs list -r pool-storage/app_storage
  NAME                              USED  AVAIL     REFER  MOUNTPOINT
  pool-storage/app_storage          430K  2.18T      151K  /pool-storage/app_storage
  pool-storage/app_storage/subfs1   140K  2.18T      140K  /pool-storage/app_storage/subfs1
  pool-storage/app_storage/subfs2   140K  2.18T      140K  /pool-storage/app_storage/subfs2  
  ```
- KVM host config:
  ```
  <filesystem type="mount" accessmode="passthrough">
  <driver type="virtiofs"/>
  <binary path="/usr/local/bin/virtiofsd-wrapper.sh" xattr="on"/>
  <source dir="/pool-storage/app_storage"/>
  <target dir="app_storage"/>
  <address type="pci" domain="0x0000" bus="0x07" slot="0x00" function="0x0"/>
  </filesystem>
  ```
- VM guest: 
  - version: Pop!_OS 22.04 LTS
  - nothing virtiofs-related in fstab
  - Mounted via cmd line `mount -t virtiofs app_storage /app_storage`


- Initially `/pool-storage/app_storage` was just a subdir on ZFS filesystem - no problem accessing files from KVM guest
- Later, I created ZFS layout as described above, files on `pool-storage/app_storage` could still be accessed, but `subfs1` and `subfs2` have problems (at least with `find`):
  ```
  # find /app_storage/
  /app_storage/
  find: File system loop detected; ‘/app_storage/subfs2’ is part of the same file system loop as ‘/app_storage/’.
  find: File system loop detected; ‘/app_storage/subfs1’ is part of the same file system loop as ‘/app_storage/’.  
```
- Despite this, I can still access files located in these `subfs*` directories
- However, despite using `--announce-submounts` (via a wrapper script, similar to Your solution) inode numbers seem to collide
- I am yet to test this with other FS, but I suspect the same problems with inode numbers to occur

**Comment by filip-paczynski** [comment]
...after chacking `mount` output on host, I noticed that I do not have `seclabel` option:
```
# mount | grep app_storage
pool-storage/app_storage on /pool-storage/app_storage type zfs (rw,xattr,noacl)
pool-storage/app_storage/subfs1 on /pool-storage/app_storage/subfs1 type zfs (rw,xattr,noacl)
pool-storage/app_storage/subfs2 on /pool-storage/app_storage/subfs2 type zfs (rw,xattr,noacl)
```

perhaps `seclabel` limits visibility of Your mountpoints?
Sadly, I know nothing about selinux.

**Comment by deajan** [comment]
@filip-paczynski Would you mind giving me the full cmdline of `virtiofsd` host side so I can compare ?

**Comment by filip-paczynski** [comment]
> @filip-paczynski Would you mind giving me the full cmdline of `virtiofsd` host side so I can compare ?

Sure:
```
# ps ax | grep virtiofs
131048 ?        S      0:00 /bin/sh /usr/local/bin/virtiofsd-wrapper.sh --fd=37 -o source=/pool-storage/app_storage,xattr
131050 ?        S      0:00  \_ /usr/lib/virtiofsd --announce-submounts --fd=37 -o source=/pool-storage/app_storage,xattr
131054 ?        Sl     0:00      \_ /usr/lib/virtiofsd --announce-submounts --fd=37 -o source=/pool-storage/app_storage,xattr
```

**Comment by filip-paczynski** [comment]
The problem with inode numbers being duplicated also occurs with XFS.
XFS layout on host side:
```
# mount | grep xfs_
/home/filip.paczynski/tmp/xfs-root on /xfs_test type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/home/filip.paczynski/tmp/xfs-subfs1 on /xfs_test/subfs1 type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
/home/filip.paczynski/tmp/xfs-subfs2 on /xfs_test/subfs2 type xfs (rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota)
```

Testing `find` on VM:
```
# mount -t virtiofs xfs_test /xfs_test/
# find /xfs_test/
/xfs_test/
find: File system loop detected; ‘/xfs_test/subfs1’ is part of the same file system loop as ‘/xfs_test/’.
find: File system loop detected; ‘/xfs_test/subfs2’ is part of the same file system loop as ‘/xfs_test/’.


```

**Comment by filip-paczynski** [comment]
I *guess* the initial issue with ZFS filesystems not being visible on VM side was related to selinux, or `seclabel` option, or similar security-related feature. 
This is no issue with ZFS.

The problem with duplicated inode numbers, which makes `find` consider `subfs*` as a hardlink to parent directory is another matter. Theoretically, such a configuration should be handled by `virtiofsd --announce-submonuts`. 
However, I do not fully understand what they mean by "device number", or how should this by handled on VM side:

from virtiofsd docs:
> --announce-submounts solves that problem because it reports a different device number for every submount it encounters.

I have no experience with `virtiofsd` and I only run KVM on my "local" machine (Xen on servers).

I guess one could ask on virtiofsd forums/discussions about `--announce-submounts` and why it duplicates inode numbers on VM side.

Glad I could help.

**Comment by deajan** [comment]
Side notes, using `xattr=on` divides IOPS by 3 on my tests.
`--inode-file-handles=never` does not resolve any of the above loop problems.

Not usable for me right now.
Thank your for your time, and sorry for having made that noise in the ZFS issues.

**Comment by zeigerpuppy** [comment]
May be related... I have noticed that virtiofs with caching enabled leaves open file handles, eventually resulting in "too many files open".  I'm not sure if this is due to an interaction with ZFS.

Disabling caching has a significant performance hit, maybe could be remedied by using the upcoming direct_io options?

https://gitlab.com/virtio-fs/virtiofsd/-/issues/121

**Comment by deajan** [comment]
For what it's worth, I found the issue.
It's not virtiofs related, but zfs related.

In my setup, when I use `zfs mount mydataset`, it mounts for the current user only.
If I happen to use `systemctl restart zfs-mount`, it mounts for all users, including the virtiofs bridge, which actually works fine.

Not sure whether this is a bug or a feature yet, but I've opened a discussion [here](https://github.com/openzfs/zfs/discussions/15208)
